design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+interfaces derive from IUnknown and all objects in any COM based system must implement these
+methods.
+
+``` cpp
+struct IUnknown
+{
+    int32_t  QueryInterface(guid const& id, void** object);
+    uint32_t AddRef();
+    uint32_t Release();
+};
+```
+
+### QueryInterface
+
+Given any COM interface pointer, you can query for any other COM interface via IUnknown's
+QueryInterface method. If the referenced COM object supports the interface being queried, a pointer
+to that COM interface is returned to the caller. Otherwise, an error code indicating that the
+requested interface is not supported.
+
+As stated above, interfaces are identified by a GUID. This GUID is passed by reference as the first
+parameter of QueryInterface. The pointer to the requested interface - if available - is returned
+via a void** out parameter. It is the responsibility of the calling code to cast this pointer to
+the correct COM interface pointer type.
+
+Methods on COM interfaces (including xlang ABI interfaces) never throw C++ exceptions, due to the
+need for a stable contract across compilers. QueryInterface returns a 32-bit signed integer as an
+error code known. On Windows, this code is known as an
+[HRESULT](https://docs.microsoft.com/en-us/windows/desktop/com/error-handling-in-com).
+Generally with HRESULTs, negative integers indicate errors, positive integers and zero indicate
+success. QueryInterface in particular should always return 0 on success.
+
+QueryInterface may return the following error codes:
+
+- 0x80004002 (E_NOINTERFACE) indicates the requested interface is not supported by this object.
+
+### AddRef and Release
+
+COM objects manage their own lifetime via an internal reference count. When this count goes to
+zero, the COM object cleans itself up. IUnknown provides methods to increment and decrement the
+object's reference count. If a language projection makes a copy of a COM interface pointer, it must
+call AddRef. When client code is finished using a given pointer to a COM interface, the language
+projection must call Release.
+
+The return value of these function is specified to be the new reference count of the object.
+However the COM standard itself warns that this value is intended to be used only for test purposes.
+In practice on Windows, the value returned from these two functions is often inaccurate. The one
+scenario where you can trust the return value is when Release returns zero - in this case, you can
+reliably assume the object has cleaned itself up.
+
+## IXlangObject
+
+In addition to the core functionality described by IUnknown, ABI interfaces for xlang types derive
+from IXlangObject. This interface adds additional core functionality needed for xlang.
+
+``` cpp
+enum ObjectInfoCategory
+{
+    StringRepresentation,
+    TypeName,
+    MemoryUsage
+};
+
+struct IXlangObject
+{
+    int32_t GetInfo(ObjectInfoCategory info_category, void** info);
```

Comment:
GetProperty?

Reply:
This is something that few people will ever type. Let's make this something a bit longer & narrower. E.g. GetObjectState or similar.

Reply:
I like making it longer but I don't like calling it object "state". How about GetObjectInfo?

Reply:
Why are we assuming it will only be used for getting?

Reply:
I'm a little worried about folks abusing this method for unintended usage as it is. Adding a set method as well (or supporting set by passing in a value via the info param) seems like asking for trouble.

Reply:
GetObjectInfo seems fine to me. For setting values, we could always add an enum value that supplies an object that then has a setter (e.g. GetObjectInfo(ResourceSettigns, &objectResourceSettings); objectResourceSettings->SetResourcePool(1);).

Reply:
I still don't like the idea of setting things via GetObjectInfo, though if nessassary I kinda like the "get a setter object" approach - I don't want that pattern to be straightforward